### DISCUSSION TITLE
Above scrollable panel image interactive problem

### DISCUSSION THREAD

**Discussion by WDL0210** [rexrainbow/phaser3-rex-notes#253] [Q&A] [answered]
First, this is my bascic game UI:
![QA00](https://user-images.githubusercontent.com/101234292/157366647-d59e8abd-767c-4c9c-9dab-d8b513b16252.png)
That have 5 image and already set interactive event.
Then i create scrollable panel via rexUI like this:
![QA01](https://user-images.githubusercontent.com/101234292/157366802-46be819f-c6f2-40ec-bfb4-0673f28111e2.png)
So...When i show a message box ui(a phaser3 container with multiple image) above scrollable panel like this:
![QA03](https://user-images.githubusercontent.com/101234292/157366994-09454d66-1404-4423-9e7a-3c1a0042c0c5.png)
The message box ui's yes and no image already add interactive event, but i find i cant' trigger my message box ui button's interavtive event.
I'm sure i can trigger button's event without create scrollable panel, it looks like scrollable panel block button's interactive.
Please tell me how to solve this problem: (

By the way, i can set [Top only](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/touchevents/#top-only) to false and then i can click my button, but that also can trigger my basic game ui's button event, so i dont' want to use this.
My phaser version are 3.55.2

Thank you reading!

**Comment by rexrainbow** [comment, marked as answer]
Since buttons on message box is placed above scrollable panel, they will receive touch event before scrollable panel ideally. 
Might try to create message box by [Dialog](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/ui-dialog/)([demo](https://codepen.io/rexrainbow/pen/MPZWZG)) instead of built-in container. Moreover, [here](https://codepen.io/rexrainbow/pen/VwzbqEP) is a [modal](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/modal-promise/) dialog.

**Reply by WDL0210** [reply]
Sounds like use modal dialog are a best way, but i still have some question, because of my game object need fit every kind screen size.
So i add all my gameobject to one container let they can auto change scale and position when screen size change or different.
But i found modal dialog can't add into Phaser3's container object, have any idea to solve?
Thank you reading.

**Reply by rexrainbow** [reply]
Any kind of rexui game object is extended from [containerlite](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/containerlite/), it can change scale and position like p3 container.

It is not suggested that putting rexui into p3 container. If you still need it, you can try `topUI.addToContainer(p3Container)`.

**Reply by rexrainbow** [reply]
topUI is a sizer which does not have parent. In this case, it might be a [dialog](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/ui-dialog/) game object.

**Reply by WDL0210** [reply]
I found a strange thing, when i use `myScrollablePanel.addToContainer(myContainer)` and `myContainer.bringToTop(myMessageBox)`.
It can solve my problem!I can trigger my message box button when above scrollable panel!
But also find some problem, when add scrollable panel to my container, the scrollable panel's child trigger will be insensitive.
By the way i use this [way ](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/ui-scrollablepanel/#set-children-interactive)to set child interactive event:
```
myScrollablePanel.setChildrenInteractive({
            target: myPanel
        }).on("child.click", myFunction, myContex);
```
So...i have a weird question, have some function can take scrollable panel out from my container?

**Reply by rexrainbow** [reply]
It might be a bug in p3's container, since you put scrollablepanel into container would solve the interactive issue.

But putting scrollablepanel into p3 container will bring other new issues, the main issue is :
**Putting game object and its mask into a container together won't work well.** See this [testing](https://codepen.io/rexrainbow/pen/mdBZJmb), enable line 22-24.

And scrollablepanel has a mask.

The other issue is, children of rexui are put into scene's display list, that's why you can't add rexui game object from p3 container directly, it only add the top game object, but not include its children. 
The same, `myContainer.bringToTop(myMessageBox)` only change the render order of top game object, not include its children.

**Reply by WDL0210** [reply]
Yeah i find my scrollable panel's mask will broke when putting into p3 container, now i know this not a good way to do...
Still need use modal dialog are best way,  back to let dailog add to container.
Now i success put dialog to my container and can auto change scale and position in  every screen size, greate :D
But sadly, the dialog's click block area(black alpha full screen image) are render lower then dialog, so i can't click dialog button.
I'm learn create dialog by this [example](https://codepen.io/rexrainbow/pen/VwzbqEP).

**Reply by rexrainbow** [reply]
> But sadly, the dialog's click block area(black alpha full screen image) are render lower then dialog, so i can't click dialog button.

Can't get it. 
Click blocking area ought to below the dialog, and above other ui elements, i.e. only the topmost dialog ui can be interactive. Which *dialog button* in this statement? Buttons on scrollable panel?

**Reply by rexrainbow** [reply]
Ah, might be related. Click blocking area is not a child of dialog, it won't be added to p3 container with dialog. 
As I said before, you can use
```javascript
dialog.setPosition( x,y ).setScale( scaleX, scaleY );
```
to change scale and position of its children.

**Reply by WDL0210** [reply]
Got it, but have some way can stop using Click blocking area?
I still prefer dialog change with container and make Click blocking area by myself.
Thank you.

**Reply by rexrainbow** [reply]
To disable cover game object feature, set `cover: false` in [config](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/modal/#create-instance) of creating modal promise.

```javascript
        this.rexUI.modalPromise(
            // Game object
            CreateDialog(this).setPosition(400, 300),
            // Config
            {
                cover: false,
                manaulClose: true,
                duration: {
                    in: 500,
                    out: 500
                }
            }
        )
```

**Reply by WDL0210** [reply]
Thanks now i can close disable cover game object and add dialog into container!
Well...now i can't click dialog's button, is button no  be add into container with dialog? like:
> Click blocking area is not a child of dialog, it won't be added to p3 container with dialog.

If it are true, i only can set dialog position and scale by my self... : (
But i just find i cant't set dialog's scale, here is my code:

My create modalPromise flow:
```javascript
this.msgb = this.CreateMessageBox().setPosition(1250, 550).setScale(0.1, 0.1);

this.scene.rexUI.modalPromise(
            // Dialog gameobject main body
            this.msgb ,
            // Config 
            {
                cover: false,
                manaulClose: true,
            }
        ).then(function (result) {
            console.log(result);
        });
```
Here is my **CreateMessageBox** function:
```javascript
CreateMessageBox() {
        this.msgBoxBG2 = this.scene.add.image(550, 250, 'mail_btn_Get').setOrigin(0);
        this.msgBoxBG2.setDisplaySize(1000, 400);
        var msgBox = this.scene.rexUI.add.dialog({
            background: this.msgBoxBG2,
            width : 1000,
            height: 400,
            content: this.scene.add.text(0, 0, 'Start Game?', {
                fontSize: '24px'
            }),

            actions: [
                this.CreateLabel('Yes'),
                this.CreateLabel('No'),
            ],

            space: {
                content: 25,
                action: 15,
      
                left: 20,
                right: 20,
                top: 20,
                bottom: 20,
            },
      
            align: {
                actions: 'right', // 'center'|'left'|'right'
            },
      
            expand: {
                content: false,  // Content is a pure text object
            }
        }).layout();

        msgBox.on('button.click', function (button, groupName, index, pointer, event) {
            msgBox.emit('modal.requestClose', {index: index, text: button.text });
        }).on('button.over', function (button, groupName, index, pointer, event) {
            button.getElement('background').setStrokeStyle(1, 0xffffff);
        }).on('button.out', function (button, groupName, index, pointer, event) {
            button.getElement('background').setStrokeStyle();
        });

        return msgBox;
    }
```
Finally, here is **CreateLabel** function:

```javascript
CreateLabel(text) {
        return this.scene.rexUI.add.label({
            background: this.scene.rexUI.add.roundRectangle(0, 0, 0, 0, 20, 0x5e92f3),

            text: this.scene.add.text(0, 0, text, {
                fontSize: '36px'
            }),
      
            space: {
                left: 50,
                right: 50,
                top: 25,
                bottom: 25
            }
        });
    }
```

**Reply by rexrainbow** [reply]
I see, it is because that ModalPromise will popUp that dialog (i.e. set scale from `0` to `1`), this behavior override your setting.

Will try to fix this scale issue later.

**Reply by WDL0210** [reply]
So can i have a way to close popUp feature?Maybe close it then i can change dialog scale?
By the way, can i have some function can tell scrollable panel stop/open input detecet?Maybe it can help me  solve my messagebox button can't interactive's problem?

**Reply by rexrainbow** [reply]
To disable transitIn/transitOut behavior, set `transitIn:false`, `transitOut:false` in config of ModalPromise
```javascript
this.scene.rexUI.modalPromise(
    // Dialog gameobject main body
   this.msgb ,
    // Config 
    {
       cover: false,
       manaulClose: true,
       transitIn: false,
       transitOut: false,
    }
```

To disable scrolling of scrollalblePanel, set `scroller: false`
```javascript
var panel = scene.rexUI.add.scrollablePanel({
    // ...
    scroller: false,
    // ...
});
```

**Reply by rexrainbow** [reply]
PopUp behavior fixed, it can popUp from `0` to **current scale value**. Please get latest version of rexui plugin.

**Reply by WDL0210** [reply]
It works!Now i can change dialog scale and disable scollable panel scrolling by `this.myScrollablePanel.scrollerEnable = false;`!
Thank you!!!

**Comment by rexrainbow** [comment]
BTW, how about using [scalemanager](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/scenemanager/), or camera zoom+scroll to do multiple screen size adaptation?

**Reply by WDL0210** [reply]
Because my team project are use too container, we not yet planned to change...
But maybe will try in next project, thanks!
